docs/guides/basic-rbac.mdx

```diff
@@ -3,267 +3,303 @@ title: Implement basic Role Based Access Control (RBAC) with metadata
+  import { checkRole } from '@/utils/roles'
+  import { clerkClient } from '@clerk/nextjs/server'
 
-## Add admin tools to find users and add roles
+  export async function setRole(formData: FormData) {
+    // Check that the user trying to set the role is an admin
+    if (!checkRole('admin')) {
+      return { message: 'Not Authorized' }
+    }
 
-You can leverage the `checkRole()` function you added along with server actions to build basic tools for finding users and managing roles.
+    try {
+      const res = await clerkClient().users.updateUser(formData.get('id') as string, {
+        publicMetadata: { role: formData.get('role') },
+      })
+      return { message: res.publicMetadata }
+    } catch (err) {
+      return { message: err }
+    }
+  }
+  ```
+
+  With the server action set up, now build the `<SearchUsers />` component. This component includes a form for searching users, and when submitted, appends the search term to the URL as a search parameter. Your `/admin` route will then perform a query based on the updated URL.
+
+  1. In your `app/admin/` directory, create a `_search-users.tsx` file.
+  1. Add the following code to the file.
+
+  ```tsx {{ filename: 'app/admin/_search-users.tsx' }}
+  'use client'
+
+  import { usePathname, useRouter } from 'next/navigation'
+
+  export const SearchUsers = () => {
+    const router = useRouter()
+    const pathname = usePathname()
+
+    return (
+      <div>
+        <form
+          onSubmit={(e) => {
+            e.preventDefault()
+            const form = e.currentTarget
+            const formData = new FormData(form)
+            const queryTerm = formData.get('search') as string
+            router.push(pathname + '?search=' + queryTerm)
+          }}
+        >
+          <label htmlFor="search">Search for Users</label>
+          <input id="search" name="search" type="text" />
+          <button type="submit">Submit</button>
+        </form>
+      </div>
+    )
+  }
+  ```
 
-Start with the server action. The `setRole()` action below will check the role of the current user using the `checkRoles()` helper to confirm that the user is an `admin`. It will then set the role for the selected user to the specified role.
+  With the server action and the search form set up, it's time to refactor the `app/admin/page.tsx` to be a server component. It will check whether a search parameter has been appended to the URL by the search form. If a search parameter is present, it will query for users matching the entered term.
 
-```ts {{ filename: 'app/admin/dashboard/_actions.ts' }}
-'use server'
+  If one or more users are found, the component will display a list of users, showing their first and last names, primary email address, and current role. Each user will have `Make Admin` and `Make Moderator` buttons, which include hidden inputs for the user ID and role. These buttons will use the `setRole()` server action to update the user's role.
```

Comment:
It might be worth adding a third button called `Clear Roles`, and a server action shockingly called `clearRoles()` which nulls the role metadata, removing any roles and returning the user to a normal user.

Reply:
[65e009e](https://github.com/clerk/clerk-docs/pull/1560/commits/65e009eba507215c97f8ed5ade542d616b465c91)

Reply:
The server action looks good, but there is no button in the page.tsx. Something like the following (untested!)

```tsx
import { setRole, clearRole } from './_actions'


<form action={clearRole}>
  <input type="hidden" value={user.id} name="id" />
  <button type="submit">Remove Role</button>
</form>
```

Reply:
Note, I used the function name as I suggested in the (`removeRole` instead of `removeRoles`). If you don't change the function name, this needs to be adjusted accordingly.

Reply:
oh I had in my code but forgot to add to the doc LOL